### ISSUE TITLE
Bad scheduling when pod/node events arrive in the wrong order

### ISSUE THREAD

**Issue by alculquicondor** [kubernetes/kubernetes#87083]
*Note that this comes out of thinking about our cache and not a user report.*

As seen in #86920, pod and node events might arrive in different order, affecting the truthfulness of our cache. Example cases are:
1. A node(s) is(are) lost. Node deletion events might arrive before their corresponding pods' deletion events. This was the focus of #86920, now fixed. This case shouldn't impact scheduling decisions, because node and all its pods will soon be gone.
2. Scheduler cold-start (more importantly, when re-starting). In this case, scheduled pods might arrive before unscheduled ones, being worst if the node creation event arrives later. This is not detrimental for most scheduling decisions, except for pod anti-affinity rules, specially if using a topology bigger than a node.

Note that we only start scheduling after informer's cached are synced. However, this check doesn't guarantee that events have been delivered to event handlers.

https://github.com/kubernetes/kubernetes/blob/21f59c99232afeae05dcdb4f2d1d156cda6a8a77/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L152-L155

Sadly, the event handler's processor is not exposed by the client in any way that would allow us to wait for a full sync before starting to schedule. **This a possible area of improvement.**

However, in HA clusters, this might not be a big issue because we keep informers' cache in sync and set the event handlers even if the instance is not the leader.

https://github.com/kubernetes/kubernetes/blob/f4e6ff84b387733e789193b74d2f46e3ea59ef01/cmd/kube-scheduler/app/server.go#L242-L261

/sig scheduling
/sig api-machinery

**Comment by alculquicondor** [comment]
@Huang-Wei, do you still have another scenario in mind related to caches?

**Comment by tedyu** [comment]
It seems we can add a func to SharedInformer :
```
        WaitForSynced(timeout int) bool
```
It would wait for Synced to become true, bounded by the given timeout.

**Comment by alculquicondor** [comment]
How is that different from `cache.WaitForCacheSync`? Unless you are referring to event handlers having synced. That's the one we could benefit from. But from my initial investigation, it didn't look trivial to do.

**Comment by tedyu** [comment]
Looking at ResourceEventHandler, there is functionality to be added which acknowledges having sync'ed.
Let me think more about this.

**Comment by Huang-Wei** [comment]
Problem 2 is an outstanding issue for non-HA environment.

> @Huang-Wei, do you still have another scenario in mind related to caches?

The issue I observed is different, I will create another issue.

**Comment by alculquicondor** [comment]
@tedyu did you find anything?
